P4RT_App_DB_Schema.md

```diff
@@ -0,0 +1,396 @@
+
+
+### Common definitions
+
+Every table entry is stored using a key made up of the table name, all match keys and the priority (if there is one). Then the value in the Redis DB is a hash map, mapping `action` to the name of the action and mapping all parameter names to the parameter values. Every key and the value in the Redis DB is represented as a JSON object that makes it consistent with the ASIC_DB representation and permitting the use of any char with variable length.
+
+**Match Fields**
+
+Match fields may be formatted as one of three types:
+
+* **Exact** (`"<value>"`)
+    * Available to all tables.
+    * Example: `"match/ether_type": "0x8000"`
+* **Ternary** (`"<value> & <mask>"`)
+    * Available for ACL tables.
+    * Example: `"match/ether_type": "0x8000 & 0xFFFF"`
+* **LPM** (`"<value>/<prefix_length>"`)
+    * Used for IP fields in FIXED_* tables (`ipv4_route` and `ipv6_route` specifically).
+    * Example: `"match/ip_6_dst": "2001:db8:85a3::/64"`
+
+Note that exact, ternary and LPM here refer not to the match kind of the match field in P4, but instead on the value/mask used:
+
+* Exact is matching on a value exactly (i.e., no mask). This is available for P4 match kinds `exact`, `optional`, and `ternary`. For `ternary`, it implies a mask of all-ones.
+* Ternary is matching on a masked value. This is only available for P4 match kinds of `ternary`.
+* LPM is matching on a value with a given prefix length. This is only available for P4 match kinds `lpm`.
+
+**Value Formatting**
+
+Values are formatted according to their type (e.g. IPv4 addresses as `10.0.0.1`). The list of formats and their description are detailed below.
+
+```
+// Describes the format of a value which has a bit width `num_bits` (unless for
+
+// the STRING format, which has no limit in the number of characters).
+enum Format {
+ // Hex string, e.g. 0x0a8b. All lowercase, and always of length
+ // ceil(num_bits/4)+2 (1 character for every 4 bits, zero-padded to be
+ // divisible by 4, and 2 characters for the '0x' prefix).
+ HEX_STRING = 0;
+ // MAC address, e.g. 00:11:ab:cd:ef:22. All lowercase, and always 17
+ // characters long.
+ MAC = 1;
+ // IPv4 address, e.g. 10.0.0.2.
+ IPV4 = 2;
+ // IPv6 address, e.g. fe80::21a:11ff:fe17:5f80. All lowercase, formatted
+ // according to RFC5952. This can be used for any num_bits of 128 or less. If
+ // num_bits is less than 128, then by convention only the upper num_bits bits
+ // can be set.
+ IPV6 = 3;
+ // String format, only printable characters.
+ STRING = 4;
+}
+```
+
+
+**Port names:**
+
+Port names refer to the ports used in SONiC and use `Format::STRING`, example “Ethernet0”.
+
+
+## Schema definition
+
+
+### APP_DB
+
+
+#### P4RT Table
+
+All data in the Application Database (APPL DB) published by the P4RT application resides in  the P4RT table. Logically, the P4RT table is divided into **subtables**:
```

Comment:
If we're planning to use generic SAI programmable objects (no P4 name in the objects) at the SAI layer, can we also use generic table names here? instead of P4RT prefix?

Reply:
In the MVP version, the P4 orchagent processes the P4 tables. In the future as the main path builds response, ownership capabilities, the P4 tables will merge into the local tables.